mitodl/mitxonline
# Optimize program enrollments API

# What are the relevant tickets?

#1700 

# Description (What does it do?)

Refactors some queries in the `ProgramSerializer` and `UserProgramEnrollmentDetailsSerializer` and `UserProgramEnrollmentsViewSet` so they should be more efficient. This also removes the implicit program enrollments from the view set - if you're not explicitly enrolled in a program, you won't see it anymore (we should have explicit enrollments currently, though). 

There are more things that could be optimized; I believe the amount of data getting pulled in through CourseSerializer is somewhat excessive too but this should result in a bit of an improvement for now. 

# How can this be tested?

As a user with enrollments in programs, load `/api/program_enrollments`. It should return reasonable data and should do so faster than the same call on the `main` branch. For best results, test each a few times sequentially. 

Additionally, the Dashboard should work as expected. Programs should appear normally; their courses should appear normally; and certificates should display as usual.
